### PR TITLE
Migration to Harmony

### LINKED ISSUE #864: patch spawn/despawn vehicle using new harmony
this code is commented out:
https://github.com/CitiesSkylinesMods/TMPE/blob/37b71da7e6f6793db1f2b579f14d96df02c7debf/TLM/TLM/Patch/_Vehicle/SpawnPatch.cs#L12
https://github.com/CitiesSkylinesMods/TMPE/blob/37b71da7e6f6793db1f2b579f14d96df02c7debf/TLM/TLM/Patch/_Vehicle/UnspawnPatch.cs#L12

Once I moved to harmony 2 should we uncomment this?

`VehicleStateManager` no longer exists but we can call `OnSpawnVehicle();` instead. that method is already being called from `CustomVehicle.CustomSpawn()` so maybe I should delete `Vehicle.Spawn/DespawnPatch`

### LINKED ISSUE #865: Use Harmony 2 instead of Redirection Framework
by doing some quick comparison I suspect much of the custom AI code is a copy of the original code with some pre/post modifications. But we have not yet moved them to harmony because of harmony1 shortcomings. but now that we have harmony 2 we should remove the custom AI redirections. The advantages are:
1- if Cities skylines change their code we will have the most up-to-date code. I am concerned that some of that code is already too old and we have not noticed.
2- easier maintenance.
3- less likely to make mistakes.

EDIT: I am dividing this issue to 3 steps:
1- migrate CustomVehicle to harmony (#864 easy peasy!)
2- migrate CustomVehicleAI to harmony (clever prefixes would do the job)
    * patch `StartPathFind()` #895
    * patch other `CustomVehicleAI` methods 
3- move CustomPathFind to harmony (this issue). This is the hardest part and requires a badass transpiler.

### LINKED ISSUE #1063: Asset Color Expander: Building color set in ACE are ignored in TMPE CustomGetColor
### Describe the problem

Hi Guys! I got back to work in Skylines and I found a incompatibility issue: The color was set in the Asset Color Expander mod is being ignore in a custom GetColor method implemented in TMPE. I pointed the code conflict below.

### Steps to reproduce

1. Subscribe and activate the [Asset Color Expander](https://steamcommunity.com/sharedfiles/filedetails/?id=2140866327)
2. Load a game
3. Go to <K> button, select the ACE tab and add a new rule
4. Type of rule filter = By Service
5. Allowed service = <Any Service>
6. Then go to second inner tab (Color settings) and change the coloring mode to Color List
7. Add a color in the list by clicking "+" button
8. It should change the color of all buildings of the city to that, but it have no effect

### Log files

N/A

### Extra appointments

I took a look in the TMPE code and found where possibly is the incompatibility source: [at TMPE/TLM/TLM/Custom/AI/CustomBuildingAI](https://github.com/CitiesSkylinesMods/TMPE/blob/3ba1f2f48ab76fc8d86f93d45c8873f345a82aac/TLM/TLM/Custom/AI/CustomBuildingAI.cs#L40)

It's applying the code of the default BuildingAI behaviour, instead of call the original method. ACE uses a pre-method patch to apply the color selection: [BuildingAIOverrides at ACE repo](https://github.com/klyte45/AssetColorExpander/blob/fecdd3a7f2629c7437224fe16c6041035fc9d77f/Overrides/ACEBuildingOverrides.cs#L24)

If there's a reason for it and I need to change ACE code to add this support, let me know and I will apply it. 

Thanks!

### PR DESCRIPTION
**Pull request by krzychu124**
#### - Improve runtime compatibility on Linux and MacOS #817 
---
- [x] Remove unused reverse redirections
- [x] Check, test and reuse Kian's work done for migrating most of `Vehicle` patches (#943, `major-feature-harmony2-migration`)
- [x] Rewrite reverse redirections to harmony delegates
- [x] Rewrite leftover RF redirections to either Harmony prefix, postfix or transpiler

---
### Improvements
- Fixed double patch causing overwriting `CarAI` redirection by `TrolleybusAI` redirection resulting in:
  - ignored `Individual driving styles`
  - ignored penalties for road conditions (studded tires, wet/damaged road)
  - possible incorrect max speed of vehicles

---
= to be continued = 
- update issues list

Closes #864  
Closes #865 
Closes #1063 

Latest build
---
You can find it [here!]( https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=improvement/full-harmony-migration) 

Compatible Node Controller:
https://github.com/kianzarrin/NodeController/releases/tag/V2.2.4

### CODE REVIEW

**Comment by kianzarrin** [comment]
> Delete RedirectionFramework from the project in favor of Harmony patching library
is this the continuation of my initiative #865 ?

**Comment by krzychu124** [comment]
> > Delete RedirectionFramework from the project in favor of Harmony patching library
> 
> is this the continuation of my initiative #865 ? might worth saying fixes 865

Yes, the plan is to fix all issues regards to harmony and finally forget about `Redirection Framework`. 
I'll try `transpiling` wherever it is possible or `prefix patch` if TM:PE changes are too big and transpiling doesn't make sense(further altering the result wouldn't be really "patchable").

**Comment by kianzarrin** [comment]
that is a load off my shoulders! thanks!

**Comment by kianzarrin** [comment]
we could use reverse patch strategy instead of using static global args struct to pass on arguments
https://discord.com/channels/131466550938042369/361891646742462467/778665308948267048
it is more thread-safe.

But that is too complicated ( I can't understand that code) and if it is hard to understand it is hard to maintain. also not sure if it will be compatible if other people also want to patch the same methods.

**Comment by krzychu124** [comment]
I'll take a look into it. One thing I found is that on Linux Harmony reverse patches tend to cause issues because it seems that mono runtime on linux is way more aggressive when it comes to method inlining. 

I'll go for easiest route to remove `Redirection Framework` first, then I'll explore potential readability/extensibility improvements. 

Personally I think it needs to work the same as before, before we will start improving/refactoring that code.

File: TLM/TLM/Patch/_VehicleAI/_AmbulanceAI/StartPathFindPatch.cs
Comment (krzychu124) [inline TLM/TLM/Patch/_VehicleAI/_AmbulanceAI/StartPathFindPatch.cs:16]: @kianzarrin did that work as intended at a time of merging?
Original method signature was `ushort vehicleId, ref Vehicle vehicleData, Vector3 startPos, Vector3 endPos, bool startBothWays, bool endBothWays, bool undergroundTarget` and in fact, we are searching for the method with such signature (using delegate description of signature) but prefixing it only with `ushort vehicleID, ref Vehicle vehicleData`.
It's similar for other `StartPathFind` prefixes. 
Am I missing something here?

**Comment by breakone9r** [comment]
Tested this last night. Seems like a step in the right direction. It solved some of the issues I've been having on Linux.

**Comment by zubozrout** [comment]
I've only played for a few hours but tried using TM:PE and loaded the game 4 times and so far so good. This work is awesome, thank you @krzychu124.

**Comment by krzychu124** [comment]
Pretty much everything was migrated, removed obsolete `RedirectionFramework` submodule.
I'll try to figure out why sometimes TM:PE patch is slightly different than stock code. Usually changes were necessary due to additional features. I need to revisit those Kian's questions in linked issues.

Mod still works, Linux users should finally use the mod without any weird crashes or issues. 
I haven't found any regression when it comes to functionality, but performance tests are welcomed.

File: TLM/TLM/Custom/PathFinding/CustomPathManager.cs
Comment (kvakvs) [inline TLM/TLM/Custom/PathFinding/CustomPathManager.cs:301]: I hope the compiler is smart enough to merge `args = default` and the following field initializations. Rider would also suggest using `args = PathCreationArgs { field = ..., field = ... }` syntax

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (kvakvs) [inline TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2420]: Style: This would read shorter with a const, something like `const pcai_conn = Constants.ManagerFactory.GameConnectionManager.PassengerCarAIConnection;`, if it accesses that const more than once, it'll even improve the performance.

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (kvakvs) [inline TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2689]: adding argument names might improve readability on this long call
plus a const for `Constants.ManagerFactory.GameConnectionManager.PassengerCarAIConnection` might help readability a little too

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (kvakvs) [inline TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2820]: consider a const for `Constants.ManagerFactory.GameConnectionManager.PassengerCarAIConnection`

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (kvakvs) [inline TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2904]: consider a const for `Constants.ManagerFactory.GameConnectionManager.PassengerCarAIConnection`

File: TLM/TLM/Manager/Impl/ExtPathManager.cs
Comment (kvakvs) [inline TLM/TLM/Manager/Impl/ExtPathManager.cs:46]: add argument names, a very long call is easier to read this way
only exception if `arg name = parameter name`, then it only adds noise

File: TLM/TLM/Patch/_TransportLineAI/StartPathFindPatch.cs
Comment (kvakvs) [inline TLM/TLM/Patch/_TransportLineAI/StartPathFindPatch.cs:20]: A super long line makes me sad 😅

File: TLM/TLM/Patch/_VehicleAI/_TrainAI/CheckNextLanePatch.cs
Comment (kvakvs) [inline TLM/TLM/Patch/_VehicleAI/_TrainAI/CheckNextLanePatch.cs:71]: multiple calls to String.Format, luckily only if logLogic is true

**Review by kvakvs — APPROVED** [review]
Nice
Only style comments and struct construction (not sure if a performance warning or not)

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (krzychu124) [reply TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2420]: Good point!

File: TLM/TLM/Manager/Impl/AdvancedParkingManager.cs
Comment (krzychu124) [reply TLM/TLM/Manager/Impl/AdvancedParkingManager.cs:2689]: Yes, I don't like such long names, so in every Harmony patch I stored references to methods to improve readability and also it's easier to compare with vanilla code if necessary. I'll change that part too 😉

File: TLM/TLM/Patch/_VehicleAI/_TrainAI/CheckNextLanePatch.cs
Comment (krzychu124) [reply TLM/TLM/Patch/_VehicleAI/_TrainAI/CheckNextLanePatch.cs:71]: Not much I can do. Usually used then debugging specific node and vehicle type - I used it only once.
One way to "solve it" is conversion to interpolated string but then it must be single line invocation - 500 chars, bad idea 😄 
I'll leave it like it is, because it's not the only place where similar problem exist.

File: TLM/TLM/Custom/PathFinding/CustomPathManager.cs
Comment (kianzarrin) [inline TLM/TLM/Custom/PathFinding/CustomPathManager.cs:278]: whats this?

Comment (krzychu124) [reply TLM/TLM/Custom/PathFinding/CustomPathManager.cs:278]: Nothing changed here, it's just some really old stuff(>2yo), 363-365 in previous version (Github is not the best in making diffs).

I moved method from 277-361 to the `ExtPathManager` (comment -> 297)

File: TLM/TLM/Custom/PathFinding/CustomPathManager.cs
Comment (kianzarrin) [inline TLM/TLM/Custom/PathFinding/CustomPathManager.cs:355]: I think vehicleTypeInfo.IsFlagSet(VehicleInfo.VehicleType.Ship) is more readable. probably not worth the effort to change it now. Just a suggestion for the future.

**Comment by kianzarrin** [comment]
> Closes #865

I am not sure if this closes #865 because for the most part we have introduced prefixes that return false.

File: TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs
Comment (kianzarrin) [inline TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:85]: If the comments about `// NON-STOCK CODE` is true, this can easily be a postfix.

Comment (kianzarrin) [reply TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:85]: its actually more difficult to have this as a prefix that returns false!

Comment (kianzarrin) [reply TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:85]: I can see you other `GetLocalisedStatusPatch`es are also using prefix so it would be nice to make it consistent.

**Review by kianzarrin — CHANGES_REQUESTED** [review]

File: TLM/TLM/Patch/_CitizenAI/_TouristAI/GetLocalizedStatusPatch.cs
Comment (krzychu124) [reply TLM/TLM/Patch/_CitizenAI/_TouristAI/GetLocalizedStatusPatch.cs:22]: Tourists don't have homes and also don't work 🤔

File: TLM/TLM/Patch/PatchCommons.cs
Comment (krzychu124) [reply TLM/TLM/Patch/PatchCommons.cs:32]: Yes, I meant target label. I know it resets assigned labels. I'll change that comment 😉 
> or maybe you meant the target label. in which case might not hurt to specify.

Why generate new exact same instruction if you can clone it?

File: TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs
Comment (krzychu124) [reply TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:25]: hehe, `HarmonyDelegate`... I'll refactor that code because `HarmonyDelegate` is creating garbage on each method call (by design probably)...

File: TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs
Comment (krzychu124) [reply TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:85]: Hmm, I'll try postfix'ing that 🤔

Comment (krzychu124) [reply TLM/TLM/Patch/_VehicleAI/_PassengerCarAI/GetLocalizedStatusPatch.cs:85]: ok... it won't be easy since we need driver instance... It's expensive to call, so I don't think it would be good idea for double call 😕 
Transpiler will be too complex, because vanilla code is set of `if` returns (aka. _labels hell_)...
I'll leave it for now. If I'll get bored, I'll try to improve it 😅

File: TLM/TLM/Util/TranspilerUtil.cs
Comment (krzychu124) [reply TLM/TLM/Util/TranspilerUtil.cs]: Changed 😉

**Review by DaEgi01 — APPROVED** [review]

File: TLM/TLM/Manager/Impl/ExtPathManager.cs
Comment (krzychu124) [reply TLM/TLM/Manager/Impl/ExtPathManager.cs:36]: Good idea!

**Review by kianzarrin — APPROVED** [review]

**Comment by kianzarrin** [comment]
@krzychu124  is `testing` branch equivalent to `major-feature-branch-harmony` ?

**Comment by kianzarrin** [comment]
when are you merging to master?

**Comment by krzychu124** [comment]
I'll wait for your changes and merge `master` to `testing` then merge `testing` back to `master` and `release/testing`  
The goal is to have `master` for continuous work(sort of changes container) and `release/testing` for workshop releases (always working version)

**Comment by DaEgi01** [comment]
so testing = labs?

**Comment by krzychu124** [comment]
Hmm, maybe "new labs", but I'll remove that branch once merged with `master`, so there will be `master` (aka development) and `release/test` and `release/stable` for easier change tracking.

**Comment by flarn2006** [comment]
I'm getting a crash when I try to toggle traffic lights for a junction while editing an intersection in the asset editor. I haven't tried it without Traffic Manager, or even with Traffic Manager on any release other than the one from this thread, but I just thought I'd mention it because it is a part of the game that I'd expect Traffic Manager to have interactions with.

**Pull request merged**
